// @nevermined-io/nevermined-sdk-js/dist/node/sdk.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.utils = exports.templates = exports.conditions = exports.Keeper = exports.Logger = exports.DID = exports.Config = exports.Account = exports.Nevermined = void 0;
const Config_1 = __importDefault(require("./models/Config"));
exports.Config = Config_1.default;
const Account_1 = __importDefault(require("./nevermined/Account"));
exports.Account = Account_1.default;
const DID_1 = __importDefault(require("./nevermined/DID"));
exports.DID = DID_1.default;
const Nevermined_1 = require("./nevermined/Nevermined");
Object.defineProperty(exports, "Nevermined", { enumerable: true, get: function () { return Nevermined_1.Nevermined; } });
const Logger_1 = require("./utils/Logger");
Object.defineProperty(exports, "Logger", { enumerable: true, get: function () { return Logger_1.LoggerInstance; } });
const Keeper_1 = __importDefault(require("./keeper/Keeper"));
exports.Keeper = Keeper_1.default;
const templates = __importStar(require("./keeper/contracts/templates"));
exports.templates = templates;
const conditions = __importStar(require("./keeper/contracts/conditions"));
exports.conditions = conditions;
const utils = __importStar(require("./utils"));
exports.utils = utils;
__exportStar(require("./ddo/DDO"), exports);
__exportStar(require("./ddo/MetaData"), exports);
var DIDRegistry_1 = require("./keeper/contracts/DIDRegistry");
Object.defineProperty(exports, "ProvenanceMethod", { enumerable: true, get: function () { return DIDRegistry_1.ProvenanceMethod; } });
var Assets_1 = require("./nevermined/Assets");
Object.defineProperty(exports, "OrderProgressStep", { enumerable: true, get: function () { return Assets_1.OrderProgressStep; } });
Object.defineProperty(exports, "CreateProgressStep", { enumerable: true, get: function () { return Assets_1.CreateProgressStep; } });
var Versions_1 = require("./nevermined/Versions");
Object.defineProperty(exports, "PlatformTechStatus", { enumerable: true, get: function () { return Versions_1.PlatformTechStatus; } });
var templates_1 = require("./keeper/contracts/templates");
Object.defineProperty(exports, "AgreementTemplate", { enumerable: true, get: function () { return templates_1.AgreementTemplate; } });
var conditions_1 = require("./keeper/contracts/conditions");
Object.defineProperty(exports, "Condition", { enumerable: true, get: function () { return conditions_1.Condition; } });
Object.defineProperty(exports, "ConditionState", { enumerable: true, get: function () { return conditions_1.ConditionState; } });
