// @nevermined-io/nevermined-sdk-js/dist/node/utils/Logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoggerInstance = exports.Logger = exports.LogLevel = void 0;
var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["None"] = -1] = "None";
    LogLevel[LogLevel["Error"] = 0] = "Error";
    LogLevel[LogLevel["Warn"] = 1] = "Warn";
    LogLevel[LogLevel["Log"] = 2] = "Log";
    LogLevel[LogLevel["Verbose"] = 3] = "Verbose";
})(LogLevel = exports.LogLevel || (exports.LogLevel = {}));
class Logger {
    constructor(logLevel = LogLevel.Verbose) {
        this.logLevel = logLevel;
    }
    setLevel(logLevel) {
        this.logLevel = logLevel;
    }
    bypass(...args) {
        this.dispatch('log', -Infinity, ...args);
    }
    debug(...args) {
        this.dispatch('debug', LogLevel.Verbose, ...args);
    }
    log(...args) {
        this.dispatch('log', LogLevel.Log, ...args);
    }
    warn(...args) {
        this.dispatch('warn', LogLevel.Warn, ...args);
    }
    error(...args) {
        this.dispatch('error', LogLevel.Error, ...args);
    }
    dispatch(verb, level, ...args) {
        if (this.logLevel >= level) {
            console[verb](...args);
        }
    }
}
exports.Logger = Logger;
exports.LoggerInstance = new Logger();
exports.default = exports.LoggerInstance;
