// @nevermined-io/nevermined-sdk-js/dist/node/models/AssetRewards.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class AssetRewards {
    constructor(..._params) {
        this.totalPrice = '0';
        this.rewards = new Map();
        if (_params.length === 1) {
            this.rewards = _params[0];
            this.rewards.forEach((v, k) => (this.totalPrice = String(Number(v) + Number(this.totalPrice))));
        }
        else if (_params.length === 2) {
            this.rewards.set(_params[0], _params[1]);
            this.totalPrice = _params[1];
        }
    }
    getTotalPrice() {
        return this.totalPrice;
    }
    getRewards() {
        return this.rewards;
    }
    getAmounts() {
        return [...this.rewards.values()];
    }
    getReceivers() {
        return [...this.rewards.keys()];
    }
    getAmountsString() {
        if (this.rewards.size == 0)
            return '[]';
        return '["' + Array.from(this.rewards.values()).join('","') + '"]';
    }
    getReceiversString() {
        if (this.rewards.size == 0)
            return '[]';
        return '["' + Array.from(this.rewards.keys()).join('","') + '"]';
    }
}
exports.default = AssetRewards;
