// @nevermined-io/nevermined-sdk-js/dist/node/keeper/contracts/templates/EscrowAccessSecretStoreTemplate.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EscrowAccessSecretStoreTemplate = void 0;
const AgreementTemplate_abstract_1 = require("./AgreementTemplate.abstract");
const BaseEscrowTemplate_abstract_1 = require("./BaseEscrowTemplate.abstract");
const utils_1 = require("../../../utils");
const EscrowAccessSecretStoreTemplate_serviceAgreementTemplate_1 = require("./EscrowAccessSecretStoreTemplate.serviceAgreementTemplate");
class EscrowAccessSecretStoreTemplate extends BaseEscrowTemplate_abstract_1.BaseEscrowTemplate {
    static getInstance(config) {
        return __awaiter(this, void 0, void 0, function* () {
            return AgreementTemplate_abstract_1.AgreementTemplate.getInstance(config, 'EscrowAccessSecretStoreTemplate', EscrowAccessSecretStoreTemplate);
        });
    }
    getServiceAgreementTemplate() {
        return __awaiter(this, void 0, void 0, function* () {
            return EscrowAccessSecretStoreTemplate_serviceAgreementTemplate_1.escrowAccessSecretStoreTemplateServiceAgreementTemplate;
        });
    }
    createAgreementFromDDO(agreementId, ddo, assetRewards, consumer, from) {
        return __awaiter(this, void 0, void 0, function* () {
            return !!(yield this.createFullAgreement(ddo.shortId(), assetRewards, consumer, from, agreementId));
        });
    }
    getAgreementIdsFromDDO(agreementId, ddo, assetRewards, consumer, from) {
        return __awaiter(this, void 0, void 0, function* () {
            const { accessSecretStoreConditionId, lockRewardConditionId, escrowRewardId } = yield this.createFullAgreementData(agreementId, ddo.shortId(), assetRewards, consumer);
            return [accessSecretStoreConditionId, lockRewardConditionId, escrowRewardId];
        });
    }
    createFullAgreement(did, assetRewards, consumer, from, agreementId = utils_1.generateId()) {
        return __awaiter(this, void 0, void 0, function* () {
            const { accessSecretStoreConditionId, lockRewardConditionId, escrowRewardId } = yield this.createFullAgreementData(agreementId, did, assetRewards, consumer);
            yield this.createAgreement(agreementId, did, [accessSecretStoreConditionId, lockRewardConditionId, escrowRewardId], [0, 0, 0], [0, 0, 0], consumer, from);
            return utils_1.zeroX(agreementId);
        });
    }
    createFullAgreementData(agreementId, did, assetRewards, consumer) {
        return __awaiter(this, void 0, void 0, function* () {
            const { didRegistry, conditions } = this.nevermined.keeper;
            const { accessSecretStoreCondition, lockRewardCondition, escrowReward } = conditions;
            const publisher = yield didRegistry.getDIDOwner(did);
            const lockRewardConditionId = yield lockRewardCondition.generateIdHash(agreementId, yield escrowReward.getAddress(), assetRewards.getTotalPrice());
            const accessSecretStoreConditionId = yield accessSecretStoreCondition.generateIdHash(agreementId, did, consumer);
            const escrowRewardId = yield escrowReward.generateIdHash(agreementId, assetRewards.getAmounts(), assetRewards.getReceivers(), publisher, consumer, lockRewardConditionId, accessSecretStoreConditionId);
            return {
                lockRewardConditionId,
                accessSecretStoreConditionId,
                escrowRewardId
            };
        });
    }
}
exports.EscrowAccessSecretStoreTemplate = EscrowAccessSecretStoreTemplate;
